docs/adr/adr-014-bootstrap-from-previous-peers.md

```diff
@@ -0,0 +1,474 @@
++       onPeersGC func([]peer.ID),
+ ) *peerTracker {
+```
+
+(_Code Snippet 1.d: Example of required changes to: go-header/p2p/peer_tracker.go_)
+
+as well as the `libhead.Exchange`'s options and construction:
+
+```diff
+ type ClientParameters struct {
+        // MaxHeadersPerRangeRequest defines the max amount of headers that can be requested per 1 request.
+        MaxHeadersPerRangeRequest uint64
+        // RangeRequestTimeout defines a timeout after which the session will try to re-request headers
+        // from another peer.
+        RangeRequestTimeout time.Duration
+        // TrustedPeersRequestTimeout a timeout for any request to a trusted peer.
+        TrustedPeersRequestTimeout time.Duration
+        // networkID is a network that will be used to create a protocol.ID
+        networkID string
+        // chainID is an identifier of the chain.
+        chainID string
++
++       OnPeersGC func([]peer.AddrInfo)
+}
+ ```
+
+(_Code Snippet 2.a: Example of required changes to: go-header/p2p/options.go_)
+
+ ```diff
+ func NewExchange[H header.Header](
+        peers peer.IDSlice,
+        connGater *conngater.BasicConnectionGater,
+        opts ...Option[ClientParameters],
+ ) (*Exchange[H], error) {
+        params := DefaultClientParameters()
+        for _, opt := range opts {
+                opt(&params)
+        }
+ 
+        err := params.Validate()
+        if err != nil {
+                return nil, err
+        }
+ 
+        ex := &Exchange[H]{
+                host:       host,
+                protocolID: protocolID(params.networkID),
+                peerTracker: newPeerTracker(
+                        host,
+                        connGater,
++                       params.OnPeersGC
+                ),
+                Params: params,
+        }
+ 
+        ex.trustedPeers = func() peer.IDSlice {
+                return shufflePeers(peers)
+        }
+        return ex, nil
+}
+ ```
+
+(_Code Snippet 2.b: Example of required changes to: go-header/p2p/exchange.go_)
+
+The event handlers to be supplied are callbacks that either:
+
+1. Put the new peer list into the peer store
+2. Remove a blocked peer from the peer store
+
+Such callbacks are easily passable as options at header module construction time, example:
+
+```diff
++   fx.Provide(newPeerStore),
+    fx.Provide(newHeaderService)
+...
+- func(cfg Config, network modp2p.Network) []p2p.Option[p2p.ClientParameters] {
++ func(cfg Config, network modp2p.Network, peerStore datastore.Datastore) []p2p.Option[p2p.ClientParameters] {
+        return []p2p.Option[p2p.ClientParameters]{
+            ...
+            p2p.WithChainID(network.String()),
++           p2p.WithOnPeersGC(func(peers []peer.ID) {
++               var topTen []peer.ID
++               if len(peers) >= 10 {
++                  topTen = peers[:9]
++               } else {
++                   topTen = peers
++               }
++               peerStore.Put(topTen)
++           }),
```

Comment:
This code snippet demonstrates where the main logic of decided Approach A is supposed to land. And it looks like the idea is to land it inside headermodule inside nodebuilder. 

This approach has two important downsides:
* go-header does not have access to the peers it stored.
	* It needs access to those peers because of [this issue](https://github.com/celestiaorg/go-header/issues/25). We want `p2p.Exchange` to be able to request on `Head`(and on other similar methods) non-bootstrap peers we persist. This issue is ___required___ to fully complete the story of coming to the network without bootstrap peers.
	* I would argue that this ADR should consider the above issue and even provide a solution. (cc @renaynay) 
* It makes the whole solution for the bootstrapping problems we have only usable by the node.
	* We can achieve greater and more impactful reusability(comparing to e.g. checkpointStore reuse I mentioned above) if we keep this logic inside of the go-header, together with any storing solution we settle on(with interface or not).
	* As of now, we have only one more consumer of go-header besides node, which is Rollkit. At a much later point, they will face the same problem as we do here. The valuable thing we can do now is to package go-header with our complete solution.

My suggestion to fix the above:
* Implement option `WithPeerPersitence(datastore.Datastore)` instead of `WithOnPeersGC` for `p2p.Exchange`
* If set, persists all the peers locally with or without GC cycles.
  * If with, we should decrease the GC cycle time to like 1 min, as 30mins are rare, and LNs running Exchange with this opt won't save anything. GCing also is not that expensive to run it that seldom.
* On start, it will auto-connect to these peers in a non-blocking manner(like we do for trusted peers in the 'p2p.Exchange')
	* Note that we cannot call these peer we persisted as trusted(will explain in further comment)

Reply:
Suggestions implemeneted 👍🏻